contracts/YieldNotionalLever.sol

```diff
@@ -0,0 +1,510 @@
+        uint256 borrowAmount,
+        uint256 fee,
+        bytes calldata data
+    ) external override returns (bytes32) {
+        Operation status = Operation(uint256(uint8(data[0])));
+        bytes6 seriesId = bytes6(data[1:7]);
+        bytes6 ilkId = bytes6(data[7:13]);
+        bytes12 vaultId = bytes12(data[13:25]);
+
+        // Test that the lender is either a fyToken contract or the join.
+        if (
+            msg.sender != address(IPool(ladle.pools(seriesId)).fyToken()) &&
+            msg.sender != address(ladle.joins(seriesId & ASSET_ID_MASK))
+        ) revert FlashLoanFailure();
+        // We trust the lender, so now we can check that we were the initiator.
+        if (initiator != address(this)) revert FlashLoanFailure();
+        // Decode the operation to execute and then call that function.
+        if (status == Operation.BORROW) {
+            uint256 baseAmount = uint256(bytes32(data[25:57]));
+            IERC20(token).approve(msg.sender, borrowAmount + fee);
+            _borrow(vaultId, seriesId, ilkId, borrowAmount, fee, baseAmount);
+        } else if (status == Operation.REPAY) {
+            IERC20(token).approve(msg.sender, borrowAmount + fee);
+            _repay(vaultId, seriesId, ilkId, borrowAmount + fee, data);
+        } else if (status == Operation.CLOSE) {
+            uint256 ink = uint256(bytes32(data[25:57]));
+            uint256 art = uint256(bytes32(data[57:89]));
+            // borrowAmount is twice as we need approval for closing the debt position & closing the flash loan
+            IERC20(token).approve(
+                msg.sender,
+                borrowAmount + borrowAmount + fee
+            );
+            _close(vaultId, ink, art);
+        }
+        return FLASH_LOAN_RETURN;
+    }
+
+    /// @notice This function is called from within the flash loan. The high
+    ///     level functionality is as follows:
+    ///         1. We have supplied dai/usdc/eth & flash loaned dai/usdc/weth
+    ///         2. We deposit it into Notional to get fCash and put it in the vault.
+    ///             2.1 In case of WETH we first convert it to ETH before depositing it into Notional
+    ///         3. Against it, we borrow enough fyDai or fyUSDC to repay the flash loan.
+    /// @param vaultId The vault id to put collateral into and borrow from.
+    /// @param seriesId The pool (and thereby series) to borrow from.
+    /// @param ilkId Id of the Ilk
+    /// @param borrowAmount The amount of DAI/USDC borrowed in the flash loan.
+    /// @param fee The fee that will be issued by the flash loan.
+    /// @param baseAmount The amount of own collateral to supply.
+    function _borrow(
+        bytes12 vaultId,
+        bytes6 seriesId,
+        bytes6 ilkId,
+        uint256 borrowAmount,
+        uint256 fee,
+        uint256 baseAmount
+    ) internal {
+        // Reuse variable to denote how much to invest. Done to prevent stack
```

Comment:
Let's not reuse parameters, instead of that move the `fCashAmount` declaration above this line, and start a block here...